ipaplatform/redhat/authconfig.py

```diff
@@ -18,15 +18,146 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+import logging
+import six
+import abc
+import re
+
 from ipaplatform.paths import paths
 from ipapython import ipautil
 from ipapython.admintool import ScriptError
 import os
 
 FILES_TO_NOT_BACKUP = ['passwd', 'group', 'shadow', 'gshadow']
 
+logger = logging.getLogger(__name__)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class RedHatAuthTool(object):
+
+    @staticmethod
+    def get_tool():
```

Comment:
What's the point of the ``get_tool`` method? The return value is hard-coded, so it doesn't serve any purpose. Just create instances of ``RedHatAuthSelect`` where you need them. If we ever need to make the code more flexible, we can easily change the other pieces.

Reply:
Is this additional call impact somehow performance? Why do you want to intricate code by spreading authselect all over the installer code for an arguable benefit?

Authselect is buggy at the moment and lack important features. We might want to replace it with something else in the future. 
Or we might want to have an installer option to allow administrator to chose from different supported pam configuration tools.

In any case it would better to stick to a bridge pattern here and logically separate pam stack configuration from the rest of the installer.

Reply:
I don't see the point of an additional layer of indirection. The abstraction layer for any caller is ``ipaplatform.tasks.tasks``. No caller outside of ``ipaplatform`` is going to use ``RedHatAuthTool``. The auth tool is an internal class.

Until the need for an additional abstraction raises and we have a clear understanding how the abstraction has to look like, the ``get_tool`` method is over-engineering. *KISS* and *YAGNI* are good guidelines.

(Also ``staticmethod`` and the location of ``get_tool`` are wrong. But let's not go there now.)

Reply:
I do see the point it keeping things separate as much as possible. 
Is it a big overhead in having a layer of abstraction on pam configuration tool?